core/corehttp/gateway_handler.go

```diff
@@ -205,14 +181,36 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
 	}
 
 	if !foundIndex {
-		// template and return directory listing
-		hndlr := webHandler{
-			"listing": dirListing,
-			"path":    urlPath,
-		}
-
 		if r.Method != "HEAD" {
-			if err := i.dirList.Execute(w, hndlr); err != nil {
+			// construct the correct back link
+			// https://github.com/ipfs/go-ipfs/issues/1365
+			var backLink string = r.URL.Path
+
+			// don't go further up than /ipfs/$hash/
+			pathSplit := strings.Split(backLink, "/")
+			switch {
```

Comment:
Only the default case here is doing something. It could be turned into a large `if !(caseA) && !(caseB) {}` but I disliked the readability.